File: app/src/main/java/com/elementary/tasks/navigation/settings/export/FragmentCloudDrives.java
package com.elementary.tasks.navigation.settings.export;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.elementary.tasks.R;
import com.elementary.tasks.core.cloud.DropboxLogin;
import com.elementary.tasks.core.cloud.GoogleLogin;
import com.elementary.tasks.core.data.AppDb;
import com.elementary.tasks.core.utils.Dialogues;
import com.elementary.tasks.core.utils.Permissions;
import com.elementary.tasks.core.utils.SuperUtil;
import com.elementary.tasks.core.views.roboto.RoboButton;
import com.elementary.tasks.databinding.FragmentCloudDrivesBinding;
import com.elementary.tasks.google_tasks.GetTaskListAsync;
import com.elementary.tasks.google_tasks.TasksCallback;
import com.elementary.tasks.navigation.settings.BaseSettingsFragment;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class FragmentCloudDrives extends BaseSettingsFragment {

    private DropboxLogin mDropbox;
    private GoogleLogin mGoogleLogin;

    private FragmentCloudDrivesBinding binding;
    private RoboButton mDropboxButton, mGoogleDriveButton;

    private ProgressDialog mDialog;
    private Handler mHandler = new Handler(Looper.getMainLooper());
    private GoogleLogin.LoginCallback mLoginCallback = new GoogleLogin.LoginCallback() {
        @Override
        public void onSuccess() {
            startSync();
        }

        @Override
        public void onFail() {
            showErrorDialog();
        }
    };
    private DropboxLogin.LoginCallback mDropboxCallback = new DropboxLogin.LoginCallback() {
        @Override
        public void onSuccess(boolean logged) {
            if (logged) {
                mDropboxButton.setText(getString(R.string.disconnect));
            } else {
                mDropboxButton.setText(getString(R.string.connect));
            }
        }
    };

    private void showErrorDialog() {
        AlertDialog.Builder builder = Dialogues.getDialog(getContext());
        builder.setMessage(getString(R.string.failed_to_login));
        builder.setPositiveButton(R.string.ok, (dialogInterface, i) -> dialogInterface.dismiss());
        builder.create().show();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentCloudDrivesBinding.inflate(inflater, container, false);
        mDropbox = new DropboxLogin(getActivity(), mDropboxCallback);
        mGoogleLogin = new GoogleLogin(getActivity(), mLoginCallback);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initDropboxButton();
        initGoogleDriveButton();
        checkGoogleStatus();
    }

    private void initGoogleDriveButton() {
        mGoogleDriveButton = binding.linkGDrive;
        mGoogleDriveButton.setOnClickListener(v -> googleDriveButtonClick());
    }

    private void googleDriveButtonClick() {
        if (Permissions.checkPermission(getActivity(),
                Permissions.GET_ACCOUNTS, Permissions.READ_EXTERNAL,
                Permissions.WRITE_EXTERNAL)) {
            switchGoogleStatus();
        } else {
            Permissions.requestPermission(getActivity(), 103,
                    Permissions.GET_ACCOUNTS, Permissions.READ_EXTERNAL,
                    Permissions.WRITE_EXTERNAL);
        }
    }

    private void initDropboxButton() {
        mDropboxButton = binding.linkDropbox;
        mDropboxButton.setOnClickListener(v -> mDropbox.login());
    }

    private void switchGoogleStatus() {
        if (!SuperUtil.checkGooglePlayServicesAvailability(getActivity())) {
            Toast.makeText(getContext(), R.string.google_play_services_not_installed, Toast.LENGTH_SHORT).show();
            return;
        }
        if (mGoogleLogin.isLogged()){
            disconnectFromGoogleServices();
        } else {
            mGoogleLogin.login();
        }
    }

    private void disconnectFromGoogleServices() {
        mGoogleLogin.logOut();
        new Thread(() -> {
            AppDb.getAppDatabase(getContext()).googleTasksDao().deleteAll();
            AppDb.getAppDatabase(getContext()).googleTaskListsDao().deleteAll();
            mHandler.post(this::finishSync);
        }).start();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (grantResults.length == 0) return;
        switch(requestCode){
            case 103:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED){
                    switchGoogleStatus();
                }
                break;
        }
    }

    private void checkGoogleStatus() {
        if (mGoogleLogin.isLogged()) {
            mGoogleDriveButton.setText(R.string.disconnect);
        } else {
            mGoogleDriveButton.setText(getString(R.string.connect));
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getCallback() != null) {
            getCallback().onTitleChange(getString(R.string.cloud_services));
            getCallback().onFragmentSelect(this);
        }
        mDropbox.checkDropboxStatus();
        checkGoogleStatus();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (mGoogleLogin != null) mGoogleLogin.onActivityResult(requestCode, resultCode, data);
    }

    private void startSync() {
        mDialog = ProgressDialog.show(getContext(), null, getString(R.string.retrieving_tasks), false, true);
        new GetTaskListAsync(getContext(), new TasksCallback() {
            @Override
            public void onFailed() {
                finishSync();
            }

            @Override
            public void onComplete() {
                finishSync();
            }
        }).execute();
    }

    private void finishSync() {
        if (mDialog != null && mDialog.isShowing()) mDialog.dismiss();
        checkGoogleStatus();
        if (getCallback() != null) getCallback().refreshMenu();
    }
}

File: app/src/main/java/com/elementary/tasks/navigation/settings/export/FragmentCloudDrives.kt
package com.elementary.tasks.navigation.settings.export

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import com.elementary.tasks.R
import com.elementary.tasks.core.app_widgets.UpdatesHelper
import com.elementary.tasks.core.cloud.DropboxLogin
import com.elementary.tasks.core.cloud.GTasks
import com.elementary.tasks.core.cloud.GoogleLogin
import com.elementary.tasks.core.cloud.storages.Dropbox
import com.elementary.tasks.core.cloud.storages.GDrive
import com.elementary.tasks.core.data.AppDb
import com.elementary.tasks.core.utils.Permissions
import com.elementary.tasks.core.utils.SuperUtil
import com.elementary.tasks.core.utils.launchDefault
import com.elementary.tasks.core.utils.withUIContext
import com.elementary.tasks.databinding.FragmentSettingsCloudDrivesBinding
import com.elementary.tasks.navigation.settings.BaseSettingsFragment
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel
import timber.log.Timber

class FragmentCloudDrives : BaseSettingsFragment<FragmentSettingsCloudDrivesBinding>() {

  private val dropbox by inject<Dropbox>()
  private val gTasks by inject<GTasks>()
  private val gDrive by inject<GDrive>()

  private val viewModel by viewModel<CloudViewModel>()
  private val mDropbox: DropboxLogin by lazy {
    DropboxLogin(requireActivity(), dropbox, mDropboxCallback)
  }
  private val mGoogleLogin: GoogleLogin by lazy {
    GoogleLogin(requireActivity(), prefs, gDrive, gTasks)
  }

  private val mDropboxCallback = object : DropboxLogin.LoginCallback {
    override fun onSuccess(b: Boolean) {
      if (b) {
        binding.linkDropbox.text = getString(R.string.disconnect)
      } else {
        binding.linkDropbox.text = getString(R.string.connect)
      }
    }
  }

  private fun showErrorDialog() {
    withContext {
      val builder = dialogues.getMaterialDialog(it)
      builder.setMessage(getString(R.string.failed_to_login))
      builder.setPositiveButton(R.string.ok) { dialogInterface, _ -> dialogInterface.dismiss() }
      builder.create().show()
    }
  }

  override fun layoutRes(): Int = R.layout.fragment_settings_cloud_drives

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    updateProgress(false)
    binding.progressMessageView.text = getString(R.string.please_wait)
    mGoogleLogin.googleStatus = {
      checkGoogleStatus()
    }
    initDropboxButton()
    initGoogleDriveButton()
    initGoogleTasksButton()

    checkGoogleStatus()
  }

  override fun onStart() {
    super.onStart()

    viewModel.isLoading.observe(viewLifecycleOwner, {
      if (it != null) {
        updateProgress(it)
      }
    })
    viewModel.isReady.observe(viewLifecycleOwner, { b ->
      if (b != null && b) {
        withContext { UpdatesHelper.updateTasksWidget(it) }
      }
    })
  }

  private fun initGoogleTasksButton() {
    withContext {
      if (SuperUtil.isGooglePlayServicesAvailable(it)) {
        binding.tasksView.visibility = View.VISIBLE
        binding.linkGTasks.setOnClickListener { googleTasksButtonClick() }
      } else {
        binding.tasksView.visibility = View.GONE
      }
    }
  }

  private fun initGoogleDriveButton() {
    withContext {
      if (SuperUtil.isGooglePlayServicesAvailable(it)) {
        binding.driveView.visibility = View.VISIBLE
        binding.linkGDrive.setOnClickListener { googleDriveButtonClick() }
      } else {
        binding.driveView.visibility = View.GONE
      }
    }
  }

  private fun googleTasksButtonClick() {
    withActivity {
      if (Permissions.checkPermission(it, 104,
          Permissions.GET_ACCOUNTS, Permissions.READ_EXTERNAL,
          Permissions.WRITE_EXTERNAL)) {
        switchGoogleTasksStatus()
      }
    }
  }

  private fun googleDriveButtonClick() {
    withActivity {
      if (Permissions.checkPermission(it, 103,
          Permissions.GET_ACCOUNTS, Permissions.READ_EXTERNAL,
          Permissions.WRITE_EXTERNAL)) {
        switchGoogleDriveStatus()
      }
    }
  }

  private fun initDropboxButton() {
    binding.linkDropbox.setOnClickListener { mDropbox.login() }
  }

  private fun switchGoogleTasksStatus() {
    withActivity {
      if (!SuperUtil.checkGooglePlayServicesAvailability(it)) {
        Toast.makeText(it, R.string.google_play_services_not_installed, Toast.LENGTH_SHORT).show()
        return@withActivity
      }
      if (mGoogleLogin.isGoogleTasksLogged) {
        disconnectFromGoogleTasks()
      } else {
        mGoogleLogin.loginTasks(object : GoogleLogin.TasksCallback {
          override fun onProgress(isLoading: Boolean) {
            updateProgress(isLoading)
          }

          override fun onResult(v: GTasks?, isLogged: Boolean) {
            Timber.d("onResult: $isLogged")
            if (isLogged) {
              viewModel.loadGoogleTasks()
            }
            checkGoogleStatus()
          }

          override fun onFail() {
            showErrorDialog()
          }
        })
      }
    }
  }

  private fun updateProgress(loading: Boolean) {
    if (loading) {
      binding.progressView.visibility = View.VISIBLE
    } else {
      binding.progressView.visibility = View.GONE
    }
    binding.linkDropbox.isEnabled = !loading
    binding.linkGDrive.isEnabled = !loading
    binding.linkGTasks.isEnabled = !loading
  }

  private fun switchGoogleDriveStatus() {
    withActivity {
      if (!SuperUtil.checkGooglePlayServicesAvailability(it)) {
        Toast.makeText(it, R.string.google_play_services_not_installed, Toast.LENGTH_SHORT).show()
        return@withActivity
      }
      if (mGoogleLogin.isGoogleDriveLogged) {
        disconnectFromGoogleDrive()
      } else {
        mGoogleLogin.loginDrive(object : GoogleLogin.DriveCallback {
          override fun onProgress(isLoading: Boolean) {
            updateProgress(isLoading)
          }

          override fun onResult(v: GDrive?, isLogged: Boolean) {
            if (isLogged) {
              checkGoogleStatus()
            }
          }

          override fun onFail() {
            showErrorDialog()
          }
        })
      }
    }
  }

  private fun disconnectFromGoogleTasks() {
    val ctx = context ?: return

    mGoogleLogin.logOutTasks()
    updateProgress(true)
    launchDefault {
      AppDb.getAppDatabase(ctx).googleTasksDao().deleteAll()
      AppDb.getAppDatabase(ctx).googleTaskListsDao().deleteAll()
      withUIContext {
        UpdatesHelper.updateTasksWidget(ctx)
        updateProgress(false)
        checkGoogleStatus()
      }
    }
  }

  private fun disconnectFromGoogleDrive() {
    mGoogleLogin.logOutDrive()
    checkGoogleStatus()
  }

  override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    if (Permissions.checkPermission(grantResults)) {
      when (requestCode) {
        103 -> switchGoogleDriveStatus()
        104 -> switchGoogleTasksStatus()
      }
    }
  }

  private fun checkGoogleStatus() {
    if (mGoogleLogin.isGoogleDriveLogged) {
      binding.linkGDrive.text = getString(R.string.disconnect)
    } else {
      binding.linkGDrive.text = getString(R.string.connect)
    }
    if (mGoogleLogin.isGoogleTasksLogged) {
      binding.linkGTasks.text = getString(R.string.disconnect)
    } else {
      binding.linkGTasks.text = getString(R.string.connect)
    }
  }

  override fun onBackStackResume() {
    super.onBackStackResume()
    mDropbox.checkDropboxStatus()
    checkGoogleStatus()
  }

  override fun onDestroy() {
    super.onDestroy()
    mGoogleLogin.googleStatus = null
  }

  override fun getTitle(): String = getString(R.string.cloud_services)

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    super.onActivityResult(requestCode, resultCode, data)
    mGoogleLogin.onActivityResult(requestCode, resultCode, data)
  }
}
